src/stream/src/executor/gap_fill.rs

```diff
@@ -0,0 +1,2000 @@
+
+/// Cache for storing Gap Fill rows.
+pub type GapFillCache = BTreeMap<GapFillCacheKey, (CompactedRow, RowType)>;
+
+const GAPFILL_CACHE_DEFAULT_CAPACITY: usize = 1024;
+
+pub struct ManagedGapFillState<S: StateStore> {
+    state_table: StateTable<S>,
+    time_key_serde: OrderedRowSerde,
+    time_column_index: usize,
+}
+
+#[derive(Clone, PartialEq, Debug)]
+pub struct GapFillStateRow {
+    pub cache_key: GapFillCacheKey,
+    pub row: OwnedRow,
+    pub row_type: RowType,
+}
+
+impl GapFillStateRow {
+    pub fn new(cache_key: GapFillCacheKey, row: OwnedRow, row_type: RowType) -> Self {
+        Self {
+            cache_key,
+            row,
+            row_type,
+        }
+    }
+}
+
+impl<S: StateStore> ManagedGapFillState<S> {
+    pub fn new(state_table: StateTable<S>, time_column_index: usize, schema: &Schema) -> Self {
+        // Create serializer for time column only.
+        let time_column_type = schema[time_column_index].data_type();
+        let time_key_serde = OrderedRowSerde::new(
+            vec![time_column_type],
+            vec![risingwave_common::util::sort_util::OrderType::ascending()],
+        );
+
+        Self {
+            state_table,
+            time_key_serde,
+            time_column_index,
+        }
+    }
+
+    pub async fn init_epoch(&mut self, epoch: EpochPair) -> StreamExecutorResult<()> {
+        self.state_table.init_epoch(epoch).await
+    }
+
+    pub fn insert(&mut self, value: impl Row) {
+        self.state_table.insert(value);
+    }
+
+    pub fn delete(&mut self, value: impl Row) {
+        self.state_table.delete(value);
+    }
+
+    /// Batch delete multiple rows from state table.
+    pub fn batch_delete(&mut self, rows: Vec<impl Row>) {
+        for row in rows {
+            self.state_table.delete(row);
+        }
+    }
+
+    /// Scans `StateStore` for filled rows between two time points (exclusive).
+    pub async fn scan_filled_rows_between(
+        &self,
+        start_time: &GapFillCacheKey,
+        end_time: &GapFillCacheKey,
+    ) -> StreamExecutorResult<Vec<(OwnedRow, OwnedRow)>> {
+        let mut filled_rows_to_delete = Vec::new();
+
+        let start_time_row = self.time_key_serde.deserialize(start_time)?;
+        let end_time_row = self.time_key_serde.deserialize(end_time)?;
+
+        let sub_range: &(Bound<OwnedRow>, Bound<OwnedRow>) = &(
+            Bound::Excluded(start_time_row),
+            Bound::Excluded(end_time_row),
+        );
+
+        let state_table_iter = self
+            .state_table
+            .iter_with_prefix(
+                &None::<row::Empty>,
+                sub_range,
+                PrefetchOptions::prefetch_for_large_range_scan(),
+            )
+            .await?;
+
+        pin_mut!(state_table_iter);
+        while let Some(row_result) = state_table_iter.next().await {
+            let row = row_result?.into_owned_row();
+
+            if let Some(is_filled_datum) = row.datum_at(row.len() - 1)
```

Comment:
I think we should know the index of `is_filled` column ahead of time, rather than compute it dynamically. We should use `self.filled_column` here.